mass_mailing_switzerland/models/mass_mailing_contact.py

```diff
@@ -16,34 +16,66 @@
 from odoo.addons.queue_job.job import job
 
 
-def _partner_split_name(partner_name):
-    return [' '.join(partner_name.split()[:-1]), ' '.join(partner_name.split()[-1:])]
-
-
-def _filter_criteria(element, partner_to_update):
-    if not partner_to_update:
-        return True
-    else:
-        if len(element) == 1:
-            element = [element]
-        return any(x.id in partner_to_update for x in element)
-
-
 class MassMailingContact(models.Model):
     _inherit = "mail.mass_mailing.contact"
 
     ##########################################################################
     #                                 FIELDS                                 #
     ##########################################################################
     partner_ids = fields.Many2many(
-        "res.partner", "mass_mailing_contact_partner_rel",
-        "partner_id", "mass_mailing_contact_id",
-        string="Associated partners", readonly=False
+        "res.partner",
+        "mass_mailing_contact_partner_rel",
+        "partner_id",
+        "mass_mailing_contact_id",
+        string="Associated partners",
+        readonly=False,
     )
+    # the principal partner is computed field from the partner ids
+    # we then use the principal partner as a stepping stone to compute the other fields
     partner_id = fields.Many2one(
-        "res.partner", "Primary partner", ondelete="cascade",
+        "res.partner",
+        string="Principal partner",
+        compute="_compute_partner_id",
+        store=True,
+        readonly=True,
+    )
+    name = fields.Char(
+        compute="_compute_name",
+        store=False,
+        readonly=True,
+    )
```

Comment:
Can be a simple related field:
```suggestion
    name = fields.Char(related="partner_id.name")
```